### DISCUSSION TITLE
[⚡] Api para envio de mensajes

### DISCUSSION THREAD

**Discussion by eduardolat** [codigoencasa/builderbot#774] [Q&A] [answered]
🚩
Antes que nada, muchas gracias por el buen trabajo al hacer esta biblioteca 🫶

### Describe tu problema
Sería muy bueno que la función `createBot` retornara un objeto con funciones para poder aprovechar la abstracción de envío de mensajes y poder reutilizar las implementaciones con los distintos proveedores que la biblioteca ya de por sí ofrece.

Esto nos permitiría extender la funcionalidad de la biblioteca haciendo nuestras propias funciones sin necesidad de crear nuestras propias abstracciones sobre los proveedores, por ejemplo, podríamos tener un chatbot y un api de WhatsApp en el mismo proyecto, podríamos configurar webhooks, enviar a clientes notificaciones de su cuenta, pedidos, etc.

Es genial como `bot-whatsapp` puede intercambiar entre proveedores, sería genial poder combinar esa función con la propuesta

### Información Adicional
Adjunto imagen con un ejemplo de como puede ser útil. No propongo agregar endpoints (¿o si?), solamente tener las funciones para poder hacerlo por cuenta propia.

![image](https://user-images.githubusercontent.com/57238554/221333129-11e5bfd5-88f2-4938-9a31-a98465d593ba.png)

**Comment by leifermendez** [comment]
@eduardolat  Buenas interesante tu enfoque afortunadamente ya puedes aprovechar esas funcionalidades te dejo un ejemplo
```js

const express = require('express')
const { createBot, createProvider, createFlow, addKeyword } = require('@bot-whatsapp/bot')

const BaileysProvider = require('@bot-whatsapp/provider/baileys')
const MockAdapter = require('@bot-whatsapp/database/mock')

const app = express()
const main = async () => {
    const adapterDB = new MockAdapter()
    const adapterFlow = createFlow([flowPrincipal])
    const adapterProvider = createProvider(BaileysProvider)

    createBot({
        flow: adapterFlow,
        provider: adapterProvider,
        database: adapterDB,
    })

    /**
     * Enviar mensaje con metodos propios del provider del bot
     */
    app.post('/send-message-bot', async (req, res) => {
        await adapterProvider.sendText('52XXXXXXXXX@c.us', 'Mensaje desde API')
        res.send({ data: 'enviado!' })
    })
    /**
     * Enviar mensajes con metodos nativos del provider
     */
    app.post('/send-message-provider', async (req, res) => {
        const id = '52XXXXXXXXX@c.us'
        const templateButtons = [
            {
                index: 1,
                urlButton: {
                    displayText: ':star: Star Baileys on GitHub!',
                    url: 'https://github.com/adiwajshing/Baileys',
                },
            },
            { index: 2, callButton: { displayText: 'Call me!', phoneNumber: '+1 (234) 5678-901' } },
            {
                index: 3,
                quickReplyButton: {
                    displayText: 'This is a reply, just like normal buttons!',
                    id: 'id-like-buttons-message',
                },
            },
        ]

        const templateMessage = {
            text: "Hi it's a template message",
            footer: 'Hello World',
            templateButtons: templateButtons,
        }

        const abc = await adapterProvider.getInstance()
        await abc.sendMessage(id, templateMessage)

        res.send({ data: 'enviado!' })
    })
    const PORT = 4000
    app.listen(PORT, () => console.log(`http://localhost:${PORT}`))
}

main()
```

**Comment by ramonbello** [comment]
Hola, @leifermendez como puedo validar que exista una session activa antes de enviar el mensaje, intente enviar el mensaje sin ninguna  session y me sale este error

![image](https://user-images.githubusercontent.com/9865312/221393838-a75da2ff-1ced-45f9-a3f3-7d440ddf4b31.png)

**Reply by RaulVzla** [reply]
Hola Ramon.. como lograste que el codigo te funcionara? yo no lo logro hacer correr.. si puedes contactame 584265714626 whatsapp o webenmerida@gmail.com

**Comment by ramonbello** [comment]
Lo solucione de esta forma:

Cree esta validación para verificar si la sesión estaba creada
 
```
async function sesionCreada() {
    try {
        const files = await readdirAsync('bot_sessions');
        return files.length !== 0;
    } catch (err) {
        console.error(err);
        return false;
    }
}
```

Luego la utilizo antes de llamar a la función del envió del mensaje:

```

app.post('/enviar-mensaje', async (req, res) => {
        const numero = req.body.numero;
        const media = req.body.url_media;
        const mensaje = req.body.mensaje;

        if (!validarCelular(numero)) {
            return res.send({
                transaccion: false,
                mensaje: 'Número de celular no válido'
            });
        }

        const existeSesion = await sesionCreada();
        if (!existeSesion) {
            return res.send({
                transaccion: false,
                mensaje: 'Sesion no creada, por favor escane el codigo qr'
            });
        }
        
        try {
            if (req.body?.hasOwnProperty('url_media')) {
                await adapterProvider.sendMedia(numero, media, mensaje);
            } else {
                await adapterProvider.sendText(numero, mensaje);
            }
            return res.send({
                transaccion: true,
                mensaje: 'enviado!'
            });

        } catch (error) {
            return res.send({
                transaccion: false,
                mensaje: 'error!'
            });
        }
    });


```

**Reply by rubenphi** [reply]
Implementé esto, pero me sigue apareciendo el siguiente mensaje de error 

```
Proveedor conectado y listo

/home/ruben/base-baileys-json/node_modules/@bot-whatsapp/provider/lib/baileys/index.cjs:596
        return this.vendor.sendMessage(number, { text: message })
                           ^

TypeError: Cannot read properties of undefined (reading 'sendMessage')
```


Intento ejecutarlo desde el main del archivo app.js, no sé si está mal. Oculto el número por seguridad. 

```
const main = async () => {
  const adapterDB = new JsonFileAdapter();
  const adapterFlow = createFlow([flowPrincipal]);
  const adapterProvider = createProvider(BaileysProvider);

  const bot = await createBot({
    flow: adapterFlow,
    provider: adapterProvider,
    database: adapterDB,
  });
  async function sesionCreada() {
    try {
      const files = await readDirAsync("bot_sessions");
      return files.length !== 0;
    } catch (err) {
      console.error(err);
      return false;
    }
  }
  const existeSesion = await sesionCreada();
  if (existeSesion) {
    const id = "************@s.whatsapp.net";

    await bot.providerClass.sendText(id, "hola este es un mensaje automatico");
  }

  QRPortalWeb();
};

main();
```

Estoy seguro que la sesión está activa, de hecho si elimino el fragmento de código en el que se llama a sendText puedo usar los flujos normalmente.

**Comment by eduardolat** [comment]
@leifermendez Muchas gracias por tu respuesta, tiene mucho sentido, ¿Crees que sería bueno documentarlo?, yo me puedo encargar de eso, tal vez en la sección, Avanzado > Providers API
Si me das el ok lo escribo y pongo un par de ejemplos

**Comment by asenawritescode** [comment]
@leifermendez 
On deployment their is an outlying issue around port allocation for the express app.

This works on local machine but fails on deployment

**Comment by mastercodelatam** [comment]
y para el archivo principal 
const main = async () => {
    //const adapterDB = new MockAdapter()
    const adapterFlow = DepenBot.createFlow([flowPrincipal]);
    const adapterProvider = DepenBot.createProvider(BaileysProvider);

    const bot = await DepenBot.createBot({
        flow: adapterFlow,
        provider: adapterProvider,
        database: adapterDB,
    })
     //await bot.providerClass.sendText('593983612878@c.us', 'hola este es un mensaje automatico');
         //await bot.providerClass.sendMedia('593983612878@c.us', 'https://Electronicos/2/6436c570e9646dd4ac88e4fc');


lo he resulto de esta manera para el archivo **Bundle.bot.cjs** 
     ```
   // 📄 Esta funcion se encarga de enviar un array de mensajes dentro de este ctx
        const sendFlow = async (messageToSend, numberOrId, options = { prev: prevMsg }) => {
            if (options.prev?.options?.capture) await cbEveryCtx(options.prev?.ref);
            const queue = [];
            for (const ctxMessage of messageToSend) {
                if (endFlowFlag) return
                const delayMs = ctxMessage?.options?.delay ?? this.generalArgs.delay ?? 0;
                if (delayMs) await delay(delayMs);
                await QueuePrincipal.enqueue(() =>
                    this.sendProviderAndSave(numberOrId, ctxMessage).then(() => resolveCbEveryCtx(ctxMessage)).catch((e)=>{
                        console.log('Ocurrio un error al obtener el cxt')
                    })
                );
            }
            return Promise.all(queue)
```

**Comment by mastercodelatam** [comment]
y para escuchar si el cliente esta listo utilice este  
```
bot.providerClass.on('ready', () => {
        console.log('Bot de WhatsApp conectado y listo');
});
```

**Comment by colcodec** [comment]
hola 
estoy tratando de responder por medio de el 
flowDynamic({  body: mensaje, from: usuario2  })
pero solo me funciona para el primer mensaje q respondi, hay alguna forma de hacer uso de el sin necesidad q intervenga en un flujo, utilizarlo fuera del flujo

**Comment by colcodec** [comment]
quisiera enviar, a cualquier numero, sin imporetar si la seccion esta activa

**Comment by sotile** [comment]
me parece super interesante esto, pero si bien tengo algunos conocimientos de javascript y php no estoy familiarizado con express y node.js, de que manera podria agregar la funcionalidad de enviar mensajes  con el bot a clientes que no iniciaron una conversacion?  hay algun ejemplo funcional que sepueda probar con postman? gracias desde ya.

**Comment by mdestafadilah** [comment]
> hola estoy tratando de responder por medio de el flowDynamic({ body: mensaje, from: usuario2 }) pero solo me funciona para el primer mensaje q respondi, hay alguna forma de hacer uso de el sin necesidad q intervenga en un flujo, utilizarlo fuera del flujo

maybe you can save the messsage to database, tthen catch by the number sender to reply again.

**Comment by mdestafadilah** [comment]
> @leifermendez Muchas gracias por tu respuesta, tiene mucho sentido, ¿Crees que sería bueno documentarlo?, yo me puedo encargar de eso, tal vez en la sección, Avanzado > Providers API Si me das el ok lo escribo y pongo un par de ejemplos

any update then?

**Comment by colcodec** [comment]
> > hola estoy tratando de responder por medio de el flowDynamic({ body: mensaje, from: usuario2 }) pero solo me funciona para el primer mensaje q respondi, hay alguna forma de hacer uso de el sin necesidad q intervenga en un flujo, utilizarlo fuera del flujo
> 
> maybe you can save the messsage to database, tthen catch by the number sender to reply again.

I am doing the same thing, I capture it, but it does not send me to the established number

**Comment by jsantacruze** [comment]
> @eduardolat Buenas interesante tu enfoque afortunadamente ya puedes aprovechar esas funcionalidades te dejo un ejemplo
> 
> ```js
> const express = require('express')
> const { createBot, createProvider, createFlow, addKeyword } = require('@bot-whatsapp/bot')
> 
> const BaileysProvider = require('@bot-whatsapp/provider/baileys')
> const MockAdapter = require('@bot-whatsapp/database/mock')
> 
> const app = express()
> const main = async () => {
>     const adapterDB = new MockAdapter()
>     const adapterFlow = createFlow([flowPrincipal])
>     const adapterProvider = createProvider(BaileysProvider)
> 
>     createBot({
>         flow: adapterFlow,
>         provider: adapterProvider,
>         database: adapterDB,
>     })
> 
>     /**
>      * Enviar mensaje con metodos propios del provider del bot
>      */
>     app.post('/send-message-bot', async (req, res) => {
>         await adapterProvider.sendText('52XXXXXXXXX@c.us', 'Mensaje desde API')
>         res.send({ data: 'enviado!' })
>     })
>     /**
>      * Enviar mensajes con metodos nativos del provider
>      */
>     app.post('/send-message-provider', async (req, res) => {
>         const id = '52XXXXXXXXX@c.us'
>         const templateButtons = [
>             {
>                 index: 1,
>                 urlButton: {
>                     displayText: ':star: Star Baileys on GitHub!',
>                     url: 'https://github.com/adiwajshing/Baileys',
>                 },
>             },
>             { index: 2, callButton: { displayText: 'Call me!', phoneNumber: '+1 (234) 5678-901' } },
>             {
>                 index: 3,
>                 quickReplyButton: {
>                     displayText: 'This is a reply, just like normal buttons!',
>                     id: 'id-like-buttons-message',
>                 },
>             },
>         ]
> 
>         const templateMessage = {
>             text: "Hi it's a template message",
>             footer: 'Hello World',
>             templateButtons: templateButtons,
>         }
> 
>         const abc = await adapterProvider.getInstance()
>         await abc.sendMessage(id, templateMessage)
> 
>         res.send({ data: 'enviado!' })
>     })
>     const PORT = 4000
>     app.listen(PORT, () => console.log(`http://localhost:${PORT}`))
> }
> 
> main()
> ```

Saludos @leifermendez he intentado replicar ese ejemplo, pero obtengo el siguiente error>
  sendTextMesage is not a function

No reconoce la función de envío.... Podrías orientarme mejor?? Gracias de antemano por la ayuda

**Comment by Nixomart** [comment]
> @eduardolat Buenas interesante tu enfoque afortunadamente ya puedes aprovechar esas funcionalidades te dejo un ejemplo
> 
> ```js
> const express = require('express')
> const { createBot, createProvider, createFlow, addKeyword } = require('@bot-whatsapp/bot')
> 
> const BaileysProvider = require('@bot-whatsapp/provider/baileys')
> const MockAdapter = require('@bot-whatsapp/database/mock')
> 
> const app = express()
> const main = async () => {
>     const adapterDB = new MockAdapter()
>     const adapterFlow = createFlow([flowPrincipal])
>     const adapterProvider = createProvider(BaileysProvider)
> 
>     createBot({
>         flow: adapterFlow,
>         provider: adapterProvider,
>         database: adapterDB,
>     })
> 
>     /**
>      * Enviar mensaje con metodos propios del provider del bot
>      */
>     app.post('/send-message-bot', async (req, res) => {
>         await adapterProvider.sendText('52XXXXXXXXX@c.us', 'Mensaje desde API')
>         res.send({ data: 'enviado!' })
>     })
>     /**
>      * Enviar mensajes con metodos nativos del provider
>      */
>     app.post('/send-message-provider', async (req, res) => {
>         const id = '52XXXXXXXXX@c.us'
>         const templateButtons = [
>             {
>                 index: 1,
>                 urlButton: {
>                     displayText: ':star: Star Baileys on GitHub!',
>                     url: 'https://github.com/adiwajshing/Baileys',
>                 },
>             },
>             { index: 2, callButton: { displayText: 'Call me!', phoneNumber: '+1 (234) 5678-901' } },
>             {
>                 index: 3,
>                 quickReplyButton: {
>                     displayText: 'This is a reply, just like normal buttons!',
>                     id: 'id-like-buttons-message',
>                 },
>             },
>         ]
> 
>         const templateMessage = {
>             text: "Hi it's a template message",
>             footer: 'Hello World',
>             templateButtons: templateButtons,
>         }
> 
>         const abc = await adapterProvider.getInstance()
>         await abc.sendMessage(id, templateMessage)
> 
>         res.send({ data: 'enviado!' })
>     })
>     const PORT = 4000
>     app.listen(PORT, () => console.log(`http://localhost:${PORT}`))
> }
> 
> main()
> ```

Como puedo hacer que esto funcione en el deploy de railway? cree las endpoint en local funciona *localhost:4000/sendMessge* pero en railwind creando un dominio e ingresando al endpoint /sendMessage me dice que no existe la pagina :/

**Comment by Nixomart** [comment]
> sendTextMesage

lo resolvi   const PORT = process.env.PORT || 4000;
  app.listen(PORT, () => console.log(`TU APP ESTA EN EL http://localhost:${PORT}`)); y en railway tenes que crear una variable PORT jaja

**Reply by matgizu** [reply]
Hice tal como lo dices pero siempre me aparece que el puerto esta en uso
le he puesto muchisismos y no funciona 
![image](https://github.com/codigoencasa/bot-whatsapp/assets/33757408/07458c71-4243-45f3-b19a-025a1d0f6d04)

**Comment by mdestafadilah** [comment]
> > @eduardolat Buenas interesante tu enfoque afortunadamente ya puedes aprovechar esas funcionalidades te dejo un ejemplo
> > ```js
> > const express = require('express')
> > const { createBot, createProvider, createFlow, addKeyword } = require('@bot-whatsapp/bot')
> > 
> > const BaileysProvider = require('@bot-whatsapp/provider/baileys')
> > const MockAdapter = require('@bot-whatsapp/database/mock')
> > 
> > const app = express()
> > const main = async () => {
> >     const adapterDB = new MockAdapter()
> >     const adapterFlow = createFlow([flowPrincipal])
> >     const adapterProvider = createProvider(BaileysProvider)
> > 
> >     createBot({
> >         flow: adapterFlow,
> >         provider: adapterProvider,
> >         database: adapterDB,
> >     })
> > 
> >     /**
> >      * Enviar mensaje con metodos propios del provider del bot
> >      */
> >     app.post('/send-message-bot', async (req, res) => {
> >         await adapterProvider.sendText('52XXXXXXXXX@c.us', 'Mensaje desde API')
> >         res.send({ data: 'enviado!' })
> >     })
> >     /**
> >      * Enviar mensajes con metodos nativos del provider
> >      */
> >     app.post('/send-message-provider', async (req, res) => {
> >         const id = '52XXXXXXXXX@c.us'
> >         const templateButtons = [
> >             {
> >                 index: 1,
> >                 urlButton: {
> >                     displayText: ':star: Star Baileys on GitHub!',
> >                     url: 'https://github.com/adiwajshing/Baileys',
> >                 },
> >             },
> >             { index: 2, callButton: { displayText: 'Call me!', phoneNumber: '+1 (234) 5678-901' } },
> >             {
> >                 index: 3,
> >                 quickReplyButton: {
> >                     displayText: 'This is a reply, just like normal buttons!',
> >                     id: 'id-like-buttons-message',
> >                 },
> >             },
> >         ]
> > 
> >         const templateMessage = {
> >             text: "Hi it's a template message",
> >             footer: 'Hello World',
> >             templateButtons: templateButtons,
> >         }
> > 
> >         const abc = await adapterProvider.getInstance()
> >         await abc.sendMessage(id, templateMessage)
> > 
> >         res.send({ data: 'enviado!' })
> >     })
> >     const PORT = 4000
> >     app.listen(PORT, () => console.log(`http://localhost:${PORT}`))
> > }
> > 
> > main()
> > ```
> 
> Como puedo hacer que esto funcione en el deploy de railway? cree las endpoint en local funciona _localhost:4000/sendMessge_ pero en railwind creando un dominio e ingresando al endpoint /sendMessage me dice que no existe la pagina :/

realy need clear documentation about this. any update then?

**Comment by sotile** [comment]
> 

@leifermendez podrias crear un tutorial para implementar lo del envio de mensajes y templates de mensajes?

**Comment by sotile** [comment]
> app.post('/enviar-mensaje', async (req, res) => {
>         const numero = req.body.numero;
>         const media = req.body.url_media;
>         const mensaje = req.body.mensaje;
> 
>         if (!validarCelular(numero)) {
>             return res.send({
>                 transaccion: false,
>                 mensaje: 'Número de celular no válido'
>             });
>         }
> 
>         const existeSesion = await sesionCreada();
>         if (!existeSesion) {
>             return res.send({
>                 transaccion: false,
>                 mensaje: 'Sesion no creada, por favor escane el codigo qr'
>             });
>         }
>         
>         try {
>             if (req.body?.hasOwnProperty('url_media')) {
>                 await adapterProvider.sendMedia(numero, media, mensaje);
>             } else {
>                 await adapterProvider.sendText(numero, mensaje);
>             }
>             return res.send({
>                 transaccion: true,
>                 mensaje: 'enviado!'
>             });
> 
>         } catch (error) {
>             return res.send({
>                 transaccion: false,
>                 mensaje: 'error!'
>             });
>         }
>     });

leifer, como recibir parametros del numero y del mensaje desde postman? aunque pase los valors en el body no logro que los tome el bot

**Reply by andrescamilo1415** [reply]
despuesde crear la app de express.... coloca las siguientes lineas.
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

**Comment by leifermendez** [comment, marked as answer]
Buenas eso sucede porque estas usando tambien QRPortal entonces se utiliza
el puerto, te comparto el repo
https://github.com/leifermendez/bot-api-railway/blob/master/app.js

**Reply by EmenetSistemas** [reply]
Basado en este ejemplo se envían urlButtons, pero esto con el sintaxis directa del proveedor, lo han intentado agregar en un addAnser..?

Algo como lo siguiente:

const flowPrincipal = addKeyword(['hola', 'menu'])
    .addAnswer(`🙌 Hola ${obtenerSaludo()}, bienvenido al chatbot de *Emenet*`)
    .addAnswer(
        [
            '🤖 ¿En que puedo ayudarte el día de hoy?',
            '',
            '1. Ver planes de internet',
            '2. Validar mi cobertura',
            '3. Conctactar con un asesor para contratar internet'
        ],
        { urlButtons: [{ body: '❌ Cancelar solicitud' }] },
        null,
        [flowConsultaPlanes, flowCoberturaInternet, flowContratacion]
    )

**Comment by carmauc** [comment]
Hola, es posible enviar un mensaje al crear un grupo con un asesor desde el bot
![Snapshot_240105140211](https://github.com/codigoencasa/bot-whatsapp/assets/111944785/d286b65c-64ec-4bfd-b96e-26d2a681cb4f)

Me continua saliendo este error TypeError: Cannot use 'in' operator to search for 'text' in Bienvenido al grupo (hKy61), aca resolveremos todas tus dudas!

**Comment by Bacelis2001** [comment]
App.post locos
